projects/js-toolkit/packages/js-toolkit-core/src/project/liferayCli/Project.ts

```diff
@@ -0,0 +1,170 @@
+/**
+ * SPDX-FileCopyrightText: © 2021 Liferay, Inc. <https://liferay.com>
+ * SPDX-License-Identifier: LGPL-3.0-or-later
+ */
+
+/* eslint-disable @liferay/no-dynamic-require */
+/* eslint-disable @typescript-eslint/no-explicit-any */
+
+import merge from 'deepmerge';
+import fs from 'fs';
+import os from 'os';
+import path from 'path';
+import resolve from 'resolve';
+
+import FilePath from '../../file/FilePath';
+import LiferayJson, {CustomElementBuildConfig} from '../../schema/LiferayJson';
+import PkgJson from '../../schema/PkgJson';
+import Build from './Build';
+import Deploy from './Deploy';
+import Dist from './Dist';
+import Start from './Start';
+
+export default class Project {
+	readonly assetsDir: FilePath | null;
+	readonly build: Build;
+	readonly deploy: Deploy;
+	readonly dir: FilePath;
+	readonly dist: Dist;
+	readonly mainModuleFile: FilePath;
+	readonly pkgJson: PkgJson;
+	readonly srcDir: FilePath;
+	readonly start: Start;
+
+	constructor(projectPath: string) {
+		this.dir = new FilePath(projectPath).resolve();
+		this.assetsDir = this.dir.join('assets');
+		this.srcDir = this.dir.join('src');
+
+		if (!fs.existsSync(this.assetsDir.asNative)) {
+			this.assetsDir = null;
+		}
+
+		this.pkgJson = JSON.parse(
+			fs.readFileSync(this.dir.join('package.json').asNative, 'utf8')
+		);
+
+		if (this.pkgJson.main) {
+			this.mainModuleFile = this.dir.join(
+				new FilePath(this.pkgJson.main, {posix: true})
+			);
+		}
+		else {
+			this.mainModuleFile = this.srcDir.join('index.js');
+		}
+
+		const liferayJson = this._loadLiferayJson();
+
+		this.build = new Build(this, liferayJson);
+		this.deploy = new Deploy(this, liferayJson);
+		this.dist = new Dist(this, liferayJson);
+		this.start = new Start(this, liferayJson);
+	}
+
+	/**
+	 * Requires a module in the context of the project (as opposed to the
+	 * context of the calling package which would just use a normal `require()`
+	 * call).
+	 * @param moduleName
+	 */
+	require(moduleName: string): any {
+		return require(this.resolve(moduleName));
+	}
+
+	/**
+	 * Resolves a module in the context of the project (as opposed to the
+	 * context of the calling package which would just use a normal
+	 * `require.resolve()` call).
+	 * @param moduleName
+	 */
+	resolve(moduleName: string): string {
+		return resolve.sync(moduleName, {
```

Comment:
Do you know if it matters if the module has a `main` file or not? I know if you use `require.resolve()`, it will error saying it can't find the module if the `main` field doesn't exist. One work around for this I have seen is resolving directly too the package.json of the module and taking the directory from there.

Reply:
I think require.resolve works in the absence of a `main` entry if there's and `index.js` file. That's the only automagic case I think...

However, this method has been copied from the bundler3 and bundler2 projects and should be agnostic to that. This is because it is a low level primitive that we need in many occasions: because we are implementing dev tools, we sometimes need to require things inside the tool itself, but many other times, we need to import things relative to the project's directory. That's what these `require` and `resolve` function in `Project` are for.

So that then you can say things like:

```javascript
   const plugin = project.require('a-babel-plugin'); // we get the plugin from user's directory
   const babel = require('babel');                           // we get babel from us, because we are providing it

   babel.use(plugin);
```

Reply:
> works in the absence of a main entry if there's and index.js file.

Oh that makes sense. I had previously seen a case where resolve was failing but it lacked both a main and an index.js